components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/handler/impl/OAuth2IntrospectionAuthenticationHandler.java

```diff
@@ -0,0 +1,160 @@
+ *
+ * WSO2 Inc. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.wso2.carbon.identity.auth.service.handler.impl;
+
+import org.apache.commons.codec.binary.Base64;
+import org.apache.commons.lang.StringUtils;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.apache.http.HttpHeaders;
+import org.wso2.carbon.identity.application.common.model.User;
+import org.wso2.carbon.identity.auth.service.AuthenticationContext;
+import org.wso2.carbon.identity.auth.service.AuthenticationRequest;
+import org.wso2.carbon.identity.auth.service.AuthenticationResult;
+import org.wso2.carbon.identity.auth.service.AuthenticationStatus;
+import org.wso2.carbon.identity.auth.service.exception.AuthClientException;
+import org.wso2.carbon.identity.auth.service.exception.AuthServerException;
+import org.wso2.carbon.identity.auth.service.exception.AuthenticationFailException;
+import org.wso2.carbon.identity.auth.service.handler.AuthenticationHandler;
+import org.wso2.carbon.identity.core.bean.context.MessageContext;
+import org.wso2.carbon.identity.core.handler.InitConfig;
+import org.wso2.carbon.identity.oauth.IdentityOAuthAdminException;
+import org.wso2.carbon.identity.oauth.common.exception.InvalidOAuthClientException;
+import org.wso2.carbon.identity.oauth.dao.OAuthConsumerDAO;
+import org.wso2.carbon.identity.oauth2.IdentityOAuth2Exception;
+import org.wso2.carbon.identity.oauth2.util.OAuth2Util;
+import org.wso2.carbon.utils.multitenancy.MultitenantUtils;
+
+import java.nio.charset.Charset;
+
+
+public class OAuth2IntrospectionAuthenticationHandler extends AuthenticationHandler {
+
+    /**
+     * OAuth2IntrospectionAuthenticationHandler is for authenticate the request based on consumer key and secret.
+     * canHandle method will confirm whether this request can be handled by this authenticator or not.
+     */
+
+    private static final Log log = LogFactory.getLog(OAuth2IntrospectionAuthenticationHandler.class);
+    private final String INTROSPECTION_URI = "oauth2/introspect";
+    private final String BASIC_AUTH_HEADER = "Basic";
+    private final String CONSUMER_KEY = "consumer-key";
+
+    @Override
+    protected AuthenticationResult doAuthenticate(MessageContext messageContext) throws AuthServerException,
+            AuthenticationFailException, AuthClientException {
+
+        BasicAuthenticationHandler basicAuthHandler = new BasicAuthenticationHandler();
```

Comment:
We don't need to explicitly call the basic authenticator here. BasicAuthenticator will get hit if this authenticator fails.

Reply:
Authenticator with the highest priority will always be called. If it fails a 401 will be returned instead of moving to other authenticators. Refer [this](https://github.com/wso2-extensions/identity-carbon-auth-rest/blob/66ad4f3c9aa4602600c55b449500b7703b41cf65/components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/AuthenticationManager.java#L97)

Reply:
Were you able to test this out and observe that behavior?